### PR TITLE
Str.__mod__  

### PR DESCRIPTION
**Pull request by jstoebel**
this implements key word arguments for `str.__mod__`

I did this by creating a function called `Args` which abstracts away the fetching of arguments.

One place I wanted to call attention is the change I made to `KeyError`. KeyError returns something like `Key Error: '\\'spam\\''` when it should just return `KeyError: 'spam'`. I'm not quite sure the best way to address this.

### CODE REVIEW

**Comment by jstoebel** [comment]
The following test is failing and I'm not sure why. I don't _think_ it has anything to do with my changes (famous last words I know). Any insights?

```
FAIL: test_subclass (tests.structures.test_class.ClassTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/stoebelj/Dropbox/batavia/tests/structures/test_class.py", line 93, in test_subclass
    """)
  File "/Users/stoebelj/Dropbox/batavia/tests/utils.py", line 444, in assertCodeExecution
    self.assertEqual(js_out, py_out, context)
AssertionError: 'Myobject instance 37 42\n37\n74\n42\n1554\nDone.\n' != 'Myobject instance 37, 42\n37\n74\n42\n1554\nDone.\n'
- Myobject instance 37 42
+ Myobject instance 37, 42
?                     +
  37
  74
  42
  1554
  Done.
 : Global context

```

**Comment by jstoebel** [comment]
Just checking in. Is there anything else this PR needs? I think its ready to go.

File: tests/datatypes/test_str.py [hunk 1]
```diff
@@ -482,6 +500,90 @@ def test_bogus_specifier(self, js_cleaner, py_cleaner):
...
+            js_bool = False,
+            decimal = False,
+            float_exp = False,
+            memory_ref = False
+        )
+        def test_wildcard_with_kwargs(self, js_cleaner, py_cleaner):
+
+            tests = adjust("""
+                print(">>> '%*(spam)s' % {'spam': 'eggs'}")
+                print('%*s' % {'spam': 'eggs'})
```
Comment (freakboy3742) [inline tests/datatypes/test_str.py]: Not sure which of these you're intending to test here, but the test case doesn't match the debug line.

File: batavia/types/StrUtils.js [hunk 2]
```diff
@@ -3,9 +3,53 @@ var type_name = require('../core').type_name;
...
+    function Args(args, usesKwargs) {
+        // represents either sequential or key word arguments.
+        // functions that need to grab arguments interface with this object
+        if (usesKwargs){
+            this.collection = args[0]
+        } else{
+            this.remainingArgs = args.slice(); // remaining args
+        }
+
+        this.getArg = function(key){
```
Comment (freakboy3742) [inline batavia/types/StrUtils.js]: Ideally this would be added to the prototype, rather than an attribute on the instance.

**Comment by jstoebel** [comment]
Thanks I'll get started on these. You mentioned formatting issues. Could you provide more guidance on what these are? You don't need to go through and mark them all, but if there are any conventions I should follow I can check for them.

**Comment by freakboy3742** [comment]
The best guidance I can give you on code formatting is to follow the general principles of spacing that is enforced by PEP8, but with 2 additions:

* if/for etc aren't functions - so `if (condition)`, not `if(condition)`
* K&R style braces (i.e., `if (condition) {`, with the closing brace at the same depth as the opening if) 

I want to set up some ESLint formatting rules so that these formatting checks can be applied automatically - but (1) I need to set that up, and (2) there's the task of fixing all the existing formatting problems in the code that I've ignored because it's too labor intensive to point them all out manually.

**Comment by jstoebel** [comment]
Thanks, I'll work on that. Just out of curiosity I ran eslint under the airbnb rules and got a BUNCH of errors some of which seem to conflict with the Python way of doing things (2 space indents for example) Should I just ignore the current eslint.rc for now?

**Comment by freakboy3742** [comment]
For the moment, yes - ignore the errors. At the moment, there's a combination of "their rules are dumb" (like 2 space indents) and actual code format problems. Hence the first problem being getting our code format style sorted.

**Comment by jstoebel** [comment]
I think I have cleaned up all of the formatting problems. Apologies in advance if I missed any!

**Comment by freakboy3742** [comment]
I am, however, going to find a way to make your editor autocorrect "percision"... :-)

**Pull request merged**